SOP-vuln/vulnerable-link-preview-extension/content_script.js

```diff
@@ -0,0 +1,121 @@
+					height : defaultHeight + "px"
+				}
+			)
+		}
+	}
+	else {
+		// Not a valid event payload
+		// This includes our own messages sent since we also catch them here it seems. 
+	}
+});
+
+// https://stackoverflow.com/questions/442404/retrieve-the-position-x-y-of-an-html-element-relative-to-the-browser-window
+// https://plainjs.com/javascript/styles/get-the-position-of-an-element-relative-to-the-document-24/
+function offset(el) {
+    var rect = el.getBoundingClientRect(),
+    scrollLeft = window.pageXOffset || document.documentElement.scrollLeft,
+    scrollTop = window.pageYOffset || document.documentElement.scrollTop;
+    return { top: rect.top + scrollTop, left: rect.left + scrollLeft }
+}
+
+async function setup() {
+	let aTags = document.getElementsByTagName("a")
+	for (let i = 0; i < aTags.length; i++) {
+		let current = aTags[i]
+		let content = await getPreviewContent(current.href)
+		if (inspectIDs.has(current.id)) {
+			postMessage({"id":current.id, "previewContent":content})
+		}
+		else {
+			pos = offset(current)
+			setPreview(
+				current, 
+				content, 
+				{
+					//left : (this.window.innerWidth / 2 - (defaultWidth / 2)) + "px", 
+					//top : (this.window.innerHeight / 2 - (defaultHeight / 2)) + "px",
+					left : (current.getBoundingClientRect().width + pos.left) + "px", 
+					top : (current.getBoundingClientRect().height + pos.top) + "px",
+					width : defaultWidth + "px", 
+					height : defaultHeight + "px"
+				}
+			)
+		}
+	}
+}
+
+function setPreview(element, content, style) {
+	// https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseenter_event
+	element.onmouseover = function() {
+		let preview = document.createElement("iframe")
+		preview.id = extensionID + "_preview_" + previewNum
+		preview.style.position = "absolute"
+		preview.style.left = style.left
+		preview.style.top = style.top
+		preview.style.width = style.width
+		preview.style.height = style.height
+		preview.style.backgroundColor = "white"
+		document.body.prepend(preview)
+		preview.contentDocument.write(content)
+		//preview.src = element.href
+	}
+	element.onmouseout = function() {
+		document.getElementById(extensionID + "_preview_" + previewNum).remove()
+	}
+	previewNum++
+}
+
+async function getPreviewContent(url) {
+	if (cachedResponses[url]) {
+		console.log("Cached response for: " + url)
+		return cachedResponses[url];
+	}
+	else {
+		return new Promise((resolve, reject) => {chrome.runtime.sendMessage(extensionID, url, function(response) {
+			console.log("Content script received content from background (so not cached response for " + url + ")")
+			cachedResponses[url] = response
+			resolve(cachedResponses[url]);
+		})});
+	}
+}
+
+// https://stackoverflow.com/questions/951021/what-is-the-javascript-version-of-sleep
+// https://stackoverflow.com/questions/28202736/how-to-execute-content-script-after-the-page-is-loaded-completely/28203168
+window.addEventListener("load", async function load(event){
```

Comment:
While I'm no expert on the usage of async, I think it just signals that the function might suspend execution. With it being a callback function it is already called asynchronously, so I don't think the async here really affects the flow of the rest of the program. I.e. we already don't know when it will execute, so saying that it might suspend execution shouldn't affect anything outside of it I think. However, I believe the keyword is necessary to allow the usage of await further down.